hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/PulsarSource.java

```diff
@@ -0,0 +1,292 @@
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    shutdownPulsarClient(pulsarClient.get());
+  }
+
+  private static Long computeTargetRecordLimit(long sourceLimit, TypedProperties props) {
+    if (sourceLimit < Long.MAX_VALUE) {
+      return sourceLimit;
+    } else {
+      return props.getLong(Config.PULSAR_SOURCE_MAX_RECORDS_PER_BATCH_THRESHOLD_PROP.key(),
+          Config.PULSAR_SOURCE_MAX_RECORDS_PER_BATCH_THRESHOLD_PROP.defaultValue());
+    }
+  }
+
+  private static String convertToOffsetString(String topic, MessageId startingOffset) {
+    return JsonUtils.topicOffsets(
+        HoodieConversionUtils.mapAsScalaImmutableMap(
+            Collections.singletonMap(topic, startingOffset)));
+  }
+
+  private static void shutdownPulsarClient(PulsarClient client) throws PulsarClientException {
+    client.close();
+    // NOTE: Current version of Pulsar's client (in Pulsar Spark Connector 3.1.1.4) is not
+    //       shutting down event-loop group properly, so we had to shut it down manually
+    try {
+      EventLoopGroup eventLoopGroup = ((PulsarClientImpl) client).eventLoopGroup();
+      if (eventLoopGroup != null) {
+        eventLoopGroup.shutdownGracefully().await();
+      }
+    } catch (InterruptedException e) {
+      // No-op
+    }
+
+    // NOTE: Pulsar clients initialized by the spark-connector, might be left not shutdown
+    //       properly (see above). To work this around we employ "nuclear" option of
+    //       fetching all Pulsar client threads and interrupting them forcibly (to make them
+    //       shutdown)
+    collectActiveThreads().stream().sequential()
+        .filter(t -> t.getName().startsWith("pulsar-client-io"))
+        .forEach(Thread::interrupt);
+  }
+
+  public static class Config {
+    private static final ConfigProperty<String> PULSAR_SOURCE_TOPIC_NAME = ConfigProperty
+        .key("hoodie.deltastreamer.source.pulsar.topic")
+        .noDefaultValue()
+        .withDocumentation("Name of the target Pulsar topic to source data from");
+
+    private static final ConfigProperty<String> PULSAR_SOURCE_SERVICE_ENDPOINT_URL = ConfigProperty
+        .key("hoodie.deltastreamer.source.pulsar.endpoint.service.url")
+        .defaultValue("pulsar://localhost:6650")
+        .withDocumentation("URL of the target Pulsar endpoint (of the form 'pulsar://host:port'");
+
+    private static final ConfigProperty<String> PULSAR_SOURCE_ADMIN_ENDPOINT_URL = ConfigProperty
+        .key("hoodie.deltastreamer.source.pulsar.endpoint.admin.url")
+        .defaultValue("http://localhost:8080")
+        .withDocumentation("URL of the target Pulsar endpoint (of the form 'pulsar://host:port'");
+
+    public enum OffsetAutoResetStrategy {
+      LATEST, EARLIEST, FAIL
+    }
+
+    private static final ConfigProperty<OffsetAutoResetStrategy> PULSAR_SOURCE_OFFSET_AUTO_RESET_STRATEGY = ConfigProperty
+        .key("hoodie.deltastreamer.source.pulsar.offset.autoResetStrategy")
```

Comment:
can we reuse the same prop `auto.offset.reset` that we use for kafka. config key is generic and I feel we can use the same for both kafka and pulsar.

Reply:
I thought about unifying the configs, but this is a much bigger project (if any) to unify configs across sources (where it makes sense).

For now i'd rather avoid coupling the sources, unless there's a very clear benefit of doing so. 

P.S. `auto.offset.reset` doesn't even follow our own naming convention for ex, b/c it's a Kafka config that we (unfortunately) share

Reply:
ok, sg.